### ISSUE TITLE
Mypy Error When using Connection.cursor method to instantiate cursor

### ISSUE THREAD

**Issue by bboggs-streambit** [laughingman7743/PyAthena#484]
<img width="1017" alt="image" src="https://github.com/laughingman7743/PyAthena/assets/126348673/8df3a231-adaf-4bc2-b3ef-48659b1a2e96">

#### (Type hinting any BaseCursor subclass results in the error):
```python
from typing import Callable
from pyathena.cursor import Cursor

class MyDbWrapperClass:
  database_cursor: Callable[[], Cursor]

  def __init__(self, database_cursor: Callable[[], Cursor] )
      self.database_cursor = database_cursor

my_db_wrapper_instance = MyDbWrapperClass(database_cursor=lambda: Connection(
                            work_group=s.ATHENA_WORKGROUP,
                            region_name=s.ATHENA_REGION,
                            cursor_class=Cursor,
                        ).cursor())
```

### Run mypy on the given file
```shell 
 mypy .
```

Result:
```shell
my-repo/my-file:linenumber error: Argument "database_cursor" to "MyDbWrapperClass" has incompatible type "Callable[[], BaseCursor]"; expected "Callable[[], Cursor]"  [arg-type]
```

## Cause of error
`BaseCursor` is type hinted as the return value of the Connection.cursor method in ``
```python
# pyathena.connection
   ....
  def cursor(self, cursor: Optional[Type[BaseCursor]] = None, **kwargs) -> BaseCursor:
```
Though every  instance of a subclass of `BaseCursor` is an instance of BaseCursor, as an abstract class, BaseCursor isn't  instantiable, and if it were, a concrete instance of the BaseCursor could not be used in place of any of the subclasses.

##### Define the TypeVar

```python
# pyathena.connection
from typing import TypeVar
...

CursorType = TypeVar("CursorType", bound=BaseCursor)
```

######  Replace BaseCursor with the type var in the type signature for `Connection.cursor` method and the type hint for the  constructor's `cursor_class` argument
```python
# pyathena.connection
....

class Connection:
    ....
    cursor_class: Type[CursorType] = Cursor,

    def cursor(self, cursor: Optional[Type[CursorType]] = None, **kwargs) -> CursorType:

```

EDIT:

It appears the solution isn't as straightforward as I thought. Digging in further....

**Comment by laughingman7743** [comment]
Pull requests are always welcome.

**Comment by laughingman7743** [comment]
I just released v3.2.0. 🎉 
https://pypi.org/project/PyAthena/3.2.0/
https://github.com/laughingman7743/PyAthena/releases/tag/v3.2.0
